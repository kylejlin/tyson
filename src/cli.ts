#! /usr/bin/env node

import { Result, result } from "rusty-ts";
import { generateTypeScriptFile, TysonConfig } from ".";
import fs from "fs";
import path from "path";
import TYSON_VERSION from "./packageVersion.generated";

type TysonCommand =
  | InvalidCommand
  | TysonHelpCommand
  | TysonVersionCommand
  | TysonGenerateCommand;

enum CommandType {
  Invalid,

  Help,
  Version,
  GenerateTypescriptFile,
}

interface InvalidCommand {
  commandType: CommandType.Invalid;
  error: InvalidTysonConfigError;
}

interface TysonHelpCommand {
  commandType: CommandType.Help;
}

interface TysonVersionCommand {
  commandType: CommandType.Version;
}

interface TysonGenerateCommand {
  commandType: CommandType.GenerateTypescriptFile;
  config: TysonConfig;
}

type InvalidTysonConfigError =
  | InvalidFileError
  | WrongTypeOfFileError
  | DuplicateGrammarFileDefsError
  | DuplicateTypeDictFileDefsError
  | DuplicateOutputFileDefsError
  | DuplicateTypeDictInterfaceDefsError
  | MissingRequiredArgsError;

enum InvalidTysonConfigErrorType {
  InvalidFile,
  WrongTypeOfFile,
  DuplicateGrammarFileDefs,
  DuplicateTypeDictFileDefs,
  DuplicateOutputFileDefs,
  DuplicateTypeDictInterfaceDefs,
  MissingRequiredArgs,
}

interface InvalidFileError {
  errorType: InvalidTysonConfigErrorType.InvalidFile;

  path: string;
}

interface WrongTypeOfFileError {
  errorType: InvalidTysonConfigErrorType.WrongTypeOfFile;

  expectedExtensions: string[];
  actualPath: string;
}

interface DuplicateGrammarFileDefsError {
  errorType: InvalidTysonConfigErrorType.DuplicateGrammarFileDefs;

  paths: [string, string];
}

interface DuplicateTypeDictFileDefsError {
  errorType: InvalidTysonConfigErrorType.DuplicateTypeDictFileDefs;

  paths: [string, string];
}

interface DuplicateOutputFileDefsError {
  errorType: InvalidTysonConfigErrorType.DuplicateOutputFileDefs;

  paths: [string, string];
}

interface DuplicateTypeDictInterfaceDefsError {
  errorType: InvalidTysonConfigErrorType.DuplicateTypeDictInterfaceDefs;

  interfaceNames: [string, string];
}

interface MissingRequiredArgsError {
  errorType: InvalidTysonConfigErrorType.MissingRequiredArgs;

  missingArgs: ("grammar-file" | "type-dict-file" | "output-file")[];
}

const HELP_TEXT = `A tool that generates a TypeScript file from your Jison grammar's semantic actions so you can typecheck those semantic actions.

USAGE:
    tyson <grammar-file> <type-dict-file> --out <output-file> [--type-dict-interface <type-dict-interface>]

OPTIONS:
    -h, --help       Prints the help info
    -v, --version    Prints the version info and exits
    
ARGUMENTS:
    <grammar-file>                      A Jison file
    <type-dict-file>                    A TypeScript file that exports a \`TysonTypeDictionary\` interface
    <output-file>                       The file to the generated code should be written to
    <type-dict-interface> (optional)    The name of the type dictionary interface. Defaults to "TysonTypeDictionary"
`;

if (require.main === module) {
  main(process.argv);
}

export function main(argv: string[]) {
  const args = argv.slice(2);
  const command = getCommand(args);
  handleCommand(command);
}

function getCommand(args: string[]): TysonCommand {
  if (args.length === 0) {
    return { commandType: CommandType.Help };
  }

  if (args[0] === "-h" || args[0] === "--help") {
    return { commandType: CommandType.Help };
  }
  if (args[0] === "-v" || args[0] === "--version") {
    return { commandType: CommandType.Version };
  }

  const config = getConfig(args);

  return config.match({
    err: (error) => {
      return { commandType: CommandType.Invalid, error };
    },

    ok: (config) => {
      return { commandType: CommandType.GenerateTypescriptFile, config };
    },
  });
}

function getConfig(
  args: string[]
): Result<TysonConfig, InvalidTysonConfigError> {
  enum AcceptMode {
    JisonAndTypeDict,
    OutputFile,
    TypeDictInterfaceName,
  }

  let mode = AcceptMode.JisonAndTypeDict;

  let pathToBnfGrammarFile: string | undefined = undefined;
  let pathToTypeDictFile: string | undefined = undefined;
  let pathToOutputFile: string | undefined = undefined;
  let typeDictInterfaceName: string | undefined = undefined;
  let emitUnusedLocations: true | undefined = undefined;
  let emitUnusedSemanticValueParams: true | undefined = undefined;

  for (const arg of args) {
    if (arg === "--in") {
      mode = AcceptMode.JisonAndTypeDict;
      continue;
    }

    if (arg === "--out" || arg === "-o") {
      mode = AcceptMode.OutputFile;
      continue;
    }

    if (arg === "--type-dict-interface" || arg === "-t") {
      mode = AcceptMode.TypeDictInterfaceName;
      continue;
    }

    if (arg === "--emit-unused-locations") {
      emitUnusedLocations = true;
      continue;
    }

    if (arg === "--emit-unused-semantic-value-params") {
      emitUnusedSemanticValueParams = true;
      continue;
    }

    if (mode === AcceptMode.JisonAndTypeDict) {
      if (/\.jison$/.test(arg)) {
        if (!isValidFile(arg)) {
          return result.err({
            errorType: InvalidTysonConfigErrorType.InvalidFile,
            path: arg,
          });
        }

        if (pathToBnfGrammarFile !== undefined) {
          return result.err({
            errorType: InvalidTysonConfigErrorType.DuplicateGrammarFileDefs,
            paths: [pathToBnfGrammarFile, arg],
          });
        }

        pathToBnfGrammarFile = arg;
        continue;
      }

      if (/\.ts$/.test(arg)) {
        if (!isValidFile(arg)) {
          return result.err({
            errorType: InvalidTysonConfigErrorType.InvalidFile,
            path: arg,
          });
        }

        if (pathToTypeDictFile !== undefined) {
          return result.err({
            errorType: InvalidTysonConfigErrorType.DuplicateTypeDictFileDefs,
            paths: [pathToTypeDictFile, arg],
          });
        }

        pathToTypeDictFile = arg;
        continue;
      }

      return result.err({
        errorType: InvalidTysonConfigErrorType.WrongTypeOfFile,
        expectedExtensions: [".jison", ".ts"],
        actualPath: arg,
      });
    }

    if (mode === AcceptMode.OutputFile) {
      if (/\.ts$/.test(arg)) {
        if (pathToOutputFile !== undefined) {
          return result.err({
            errorType: InvalidTysonConfigErrorType.DuplicateOutputFileDefs,
            paths: [pathToOutputFile, arg],
          });
        }

        pathToOutputFile = arg;
        continue;
      }

      return result.err({
        errorType: InvalidTysonConfigErrorType.WrongTypeOfFile,
        expectedExtensions: [".ts"],
        actualPath: arg,
      });
    }

    if (mode === AcceptMode.TypeDictInterfaceName) {
      if (typeDictInterfaceName !== undefined) {
        return result.err({
          errorType: InvalidTysonConfigErrorType.DuplicateTypeDictInterfaceDefs,
          interfaceNames: [typeDictInterfaceName, arg],
        });
      }

      typeDictInterfaceName = arg;
      continue;
    }
  }

  if (
    pathToBnfGrammarFile !== undefined &&
    pathToTypeDictFile !== undefined &&
    pathToOutputFile !== undefined
  ) {
    return result.ok({
      pathToBnfGrammarFile,
      pathToTypeDictFile,
      pathToOutputFile,
      typeDictInterfaceName,
      emitUnusedLocations,
      emitUnusedSemanticValueParams,
    });
  } else {
    return result.err({
      errorType: InvalidTysonConfigErrorType.MissingRequiredArgs,
      missingArgs: [
        ...((pathToBnfGrammarFile === undefined ? ["grammar-file"] : []) as
          | ["grammar-file"]
          | []),
        ...((pathToTypeDictFile === undefined ? ["type-dict-file"] : []) as
          | ["type-dict-file"]
          | []),
        ...((pathToOutputFile === undefined ? ["output-file"] : []) as
          | ["output-file"]
          | []),
      ],
    });
  }
}

function isValidFile(p: string): boolean {
  const absolutePath = path.isAbsolute(p) ? p : path.join(process.cwd(), p);
  return fs.existsSync(absolutePath) && fs.lstatSync(absolutePath).isFile();
}

function handleCommand(command: TysonCommand): void {
  switch (command.commandType) {
    case CommandType.Invalid:
      handleInvalidCommand(command);
      break;
    case CommandType.Help:
      handleHelpCommand();
      break;
    case CommandType.Version:
      handleVersionCommand();
      break;
    case CommandType.GenerateTypescriptFile:
      handleGenerateCommand(command);
      break;
  }
}

function handleInvalidCommand(command: InvalidCommand): void {
  const { error } = command;
  switch (error.errorType) {
    case InvalidTysonConfigErrorType.InvalidFile:
      handleInvalidFileError(error);
      break;
    case InvalidTysonConfigErrorType.WrongTypeOfFile:
      handleWrongTypeOfFileError(error);
      break;
    case InvalidTysonConfigErrorType.DuplicateGrammarFileDefs:
      handleDuplicateGrammarFileDefsError(error);
      break;
    case InvalidTysonConfigErrorType.DuplicateTypeDictFileDefs:
      handleDuplicateTypeDictFileDefsError(error);
      break;
    case InvalidTysonConfigErrorType.DuplicateOutputFileDefs:
      handleDuplicateOutputFileDefsError(error);
      break;
    case InvalidTysonConfigErrorType.DuplicateTypeDictInterfaceDefs:
      handleDuplicateTypeDictInterfaceDefsError(error);
      break;
    case InvalidTysonConfigErrorType.MissingRequiredArgs:
      handleMissingRequiredArgsError(error);
      break;
  }
}

function handleInvalidFileError(error: InvalidFileError): void {
  console.log('No file named "' + error.path + '" exists.');
  process.exit(1);
}

function handleWrongTypeOfFileError(error: WrongTypeOfFileError): void {
  console.log(
    "Expecting a file with one of the following extensions: [" +
      error.expectedExtensions.map((ext) => '"' + ext + '"').join(", ") +
      '], but received "' +
      error.actualPath +
      '".'
  );
  process.exit(1);
}

function handleDuplicateGrammarFileDefsError(
  error: DuplicateGrammarFileDefsError
): void {
  console.log(
    'Duplicate grammar file definitions:\n    First:  "' +
      error.paths[0] +
      '"\n    Second: "' +
      error.paths[1] +
      '"'
  );
}

function handleDuplicateTypeDictFileDefsError(
  error: DuplicateTypeDictFileDefsError
): void {
  console.log(
    'Duplicate type dictionary file definitions:\n    First:  "' +
      error.paths[0] +
      '"\n    Second: "' +
      error.paths[1] +
      '"'
  );
}

function handleDuplicateOutputFileDefsError(
  error: DuplicateOutputFileDefsError
): void {
  console.log(
    'Duplicate output file definitions:\n    First:  "' +
      error.paths[0] +
      '"\n    Second: "' +
      error.paths[1] +
      '"'
  );
}

function handleDuplicateTypeDictInterfaceDefsError(
  error: DuplicateTypeDictInterfaceDefsError
): void {
  console.log(
    'Duplicate type dictionary interface definitions:\n    First:  "' +
      error.interfaceNames[0] +
      '"\n    Second: "' +
      error.interfaceNames[1] +
      '"'
  );
}

function handleMissingRequiredArgsError(error: MissingRequiredArgsError): void {
  console.log(
    "Missing the following required arguments: [" +
      error.missingArgs.join(", ") +
      "]."
  );
  process.exit(1);
}

function handleHelpCommand(): void {
  console.log(HELP_TEXT);
}

function handleVersionCommand(): void {
  console.log("tyson " + TYSON_VERSION);
}

function handleGenerateCommand(command: TysonGenerateCommand): void {
  generateTypeScriptFile(command.config);
}
