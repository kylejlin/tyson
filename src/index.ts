import fs from "fs";
import path from "path";
import prettier from "prettier";
import bnfParser, { ParsedBnfGrammar } from "./lib/bnf-parser";

export interface TysonConfig {
  pathToBnfGrammarFile: string;
  pathToTypeDictFile: string;
  pathToOutputFile: string;
  typeDictInterfaceName?: string;
}

const DEFAULT_TYPE_DICT_INTERFACE_NAME = "TysonTypeDictionary";

export function generateTypeScriptFile(config: TysonConfig): void {
  const { pathToBnfGrammarFile, pathToOutputFile } = config;
  const grammarSource = fs.readFileSync(pathToBnfGrammarFile, "utf8");
  const parsedGrammar = bnfParser.parse(grammarSource);

  const generatedCode = generateCode(config, parsedGrammar);

  const prettierConfig =
    prettier.resolveConfig.sync(config.pathToOutputFile) || {};

  fs.writeFileSync(
    pathToOutputFile,
    prettier.format(generatedCode, {
      ...prettierConfig,
      parser: "typescript",
    })
  );
}

function generateCode(config: TysonConfig, grammar: ParsedBnfGrammar): string {
  const { start } = grammar;

  if (start === undefined) {
    throw new Error(
      "The grammar defined in " +
        config.pathToBnfGrammarFile +
        " does not have a start symbol declared. Please declare one using `%start <symbolName>`."
    );
  }

  const dependencies = getActionDependencies(grammar);

  const importStatement = generateImportStatementCode(config, dependencies);
  const yyDeclarations = generateYyDeclarations(dependencies);
  const locationInterfaceDeclaration = generateTokenLocationInterfaceDeclarationIfNeeded(
    dependencies
  );
  const semanticActions = generateSemanticActionCode(
    {
      ...grammar,
      start,
    },
    getTypeDictInterfaceName(config)
  );

  return (
    importStatement +
    "\n\n" +
    yyDeclarations +
    "\n\n" +
    locationInterfaceDeclaration +
    "\n\n" +
    semanticActions
  );
}

function generateImportStatementCode(
  config: TysonConfig,
  dependencies: ActionDependencies
): string {
  const relativePath = path.relative(
    path.dirname(config.pathToOutputFile),
    config.pathToTypeDictFile
  );
  const withoutExtension = relativePath.replace(/(?:\.d)?\.ts/, "");
  const withLeadingDotSlash =
    withoutExtension.charAt(0) === "."
      ? withoutExtension
      : "." + path.sep + withoutExtension;

  return (
    "import { " +
    (dependencies.yy ? "yy, " : "") +
    getTypeDictInterfaceName(config) +
    " } from " +
    JSON.stringify(withLeadingDotSlash) +
    ";"
  );
}

function getTypeDictInterfaceName(config: TysonConfig): string {
  if (config.typeDictInterfaceName === undefined) {
    return DEFAULT_TYPE_DICT_INTERFACE_NAME;
  }

  return config.typeDictInterfaceName;
}

function generateYyDeclarations(dependencies: ActionDependencies): string {
  const yytext = dependencies.yytext ? "declare var yytext: string;" : "";
  const yyleng = dependencies.yyleng ? "declare var yyleng: string;" : "";
  const yylineno = dependencies.yylineno ? "declare var yylineno: string;" : "";
  return yytext + yyleng + yylineno;
}

function getActionDependencies(grammar: ParsedBnfGrammar): ActionDependencies {
  let tokenLocation = false;
  let yy = false;
  let yytext = false;
  let yyleng = false;
  let yylineno = false;

  Object.values(grammar.bnf).forEach((ruleActionPairs) =>
    ruleActionPairs.map(normalizeRuleActionPair).forEach(({ action }) => {
      if (/@(?:\$|-?\d+\b)/.test(action)) {
        tokenLocation = true;
      }

      if (/\byy\b/.test(action)) {
        yy = true;
      }

      if (/\byytext\b/.test(action)) {
        yytext = true;
      }

      if (/\byyleng\b/.test(action)) {
        yyleng = true;
      }

      if (/\byylineno\b/.test(action)) {
        yylineno = true;
      }
    })
  );

  return { tokenLocation, yy, yytext, yyleng, yylineno };
}

interface ActionDependencies {
  tokenLocation: boolean;
  yy: boolean;
  yytext: boolean;
  yyleng: boolean;
  yylineno: boolean;
}

function generateTokenLocationInterfaceDeclarationIfNeeded(
  dependencies: ActionDependencies
): string {
  return dependencies.tokenLocation
    ? "interface TokenLocation { first_line: number; last_line: number; first_column: number; last_column: number; range: [number, number]; }"
    : "";
}

interface ParsedBnfGrammarWithStartSymbol extends ParsedBnfGrammar {
  start: string;
}

function generateSemanticActionCode(
  grammar: ParsedBnfGrammarWithStartSymbol,
  typeDictInterfaceName: string
): string {
  const { start: startSymbol, bnf } = grammar;

  let code = "const semanticActions = {";

  Object.entries(bnf).forEach(([symbolName, ruleActionPairs]) => {
    ruleActionPairs.forEach((ruleActionPair) => {
      const { rule, action } = normalizeRuleActionPair(ruleActionPair);
      const methodDeclaration =
        JSON.stringify(symbolName + " -> " + rule) +
        getParenthesizedArgs(rule, action, typeDictInterfaceName) +
        getReturnAnnotation(symbolName, typeDictInterfaceName) +
        "{" +
        get$$Declaration(symbolName, typeDictInterfaceName) +
        action.replace(/@(\$|-?\d+\b)/g, 'yylstack["@$1"]') +
        (symbolName === startSymbol ? "}," : "\nreturn $$;},");
      code += methodDeclaration + "\n\n";
    });
  });

  code += "};";

  return code;
}

function normalizeRuleActionPair(
  ruleActionPair:
    | string
    | [string, string]
    | [string, { prec: string }]
    | [string, string, { prec: string }]
): { rule: string; action: string } {
  if (typeof ruleActionPair === "string") {
    return { rule: ruleActionPair, action: "$$ = $1;" };
  } else if ("object" === typeof ruleActionPair[1]) {
    return { rule: ruleActionPair[0], action: "$$ = $1;" };
  } else {
    return { rule: ruleActionPair[0], action: ruleActionPair[1] };
  }
}

function getParenthesizedArgs(
  rule: string,
  action: string,
  typeDictInterfaceName: string
): string {
  const actionReferencesTokenLocation = /@(\$|-?\d+\b)/g.test(action);
  const possibleYylstackArgDefWithComma = actionReferencesTokenLocation
    ? "yylstack: " + getYylstackType(rule) + ",\n\n"
    : "";
  return (
    "(" +
    possibleYylstackArgDefWithComma +
    rule
      .split(/\s+/g)
      .map(
        (arg, i) =>
          "$" +
          (i + 1) +
          ": " +
          typeDictInterfaceName +
          "[" +
          JSON.stringify(arg) +
          "]"
      )
      .join(", ") +
    ")"
  );
}

function getYylstackType(rule: string): string {
  return (
    "{" +
    ["@$"]
      .concat(rule.split(/\s+/g).map((_arg, i) => "@" + (i + 1)))
      .map((key) => JSON.stringify(key) + ": TokenLocation")
      .join(", ") +
    "}"
  );
}

function getReturnAnnotation(
  symbolName: string,
  typeDictInterfaceName: string
): string {
  return ": " + typeDictInterfaceName + "[" + JSON.stringify(symbolName) + "]";
}

function get$$Declaration(
  symbolName: string,
  typeDictInterfaceName: string
): string {
  return (
    "let $$: " + typeDictInterfaceName + "[" + JSON.stringify(symbolName) + "];"
  );
}
