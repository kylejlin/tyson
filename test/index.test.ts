import fs from "fs";
import path from "path";
import rimraf from "rimraf";
import { generateTypeScriptFile } from "../src/";
import { getErrorDiagnostics } from "./util/compiler";

beforeAll(() => {
  rimraf.sync(path.join(__dirname, "./tempIndexTest"));
  fs.mkdirSync(path.join(__dirname, "./tempIndexTest"));
  fs.mkdirSync(path.join(__dirname, "./tempIndexTest/noTypeErrors"));
  fs.mkdirSync(path.join(__dirname, "./tempIndexTest/typeErrors"));
});

test("success cases", () => {
  const dirs = fs.readdirSync(path.join(__dirname, "./fixtures/shouldSucceed"));

  dirs.forEach((dir) => {
    const pathToBnfGrammarFile = path.join(
      __dirname,
      "fixtures/shouldSucceed/" + dir + "/" + dir + ".jison"
    );
    const pathToTypeDictFile = path.join(
      __dirname,
      "fixtures/shouldSucceed/" + dir + "/" + dir + ".ts"
    );
    const pathToOutputFile = path.join(
      __dirname,
      "./tempIndexTest/noTypeErrors/" + dir + ".generated.ts"
    );

    generateTypeScriptFile({
      pathToBnfGrammarFile,
      pathToTypeDictFile,
      pathToOutputFile,
    });

    expect(getErrorDiagnostics([pathToOutputFile])).toHaveLength(0);
  });
});

test("failure cases", () => {
  const dirs = fs.readdirSync(path.join(__dirname, "./fixtures/shouldFail"));

  dirs.forEach((dir) => {
    const pathToBnfGrammarFile = path.join(
      __dirname,
      "fixtures/shouldFail/" + dir + "/" + dir + ".jison"
    );
    const pathToTypeDictFile = path.join(
      __dirname,
      "fixtures/shouldFail/" + dir + "/" + dir + ".ts"
    );
    const pathToOutputFile = path.join(
      __dirname,
      "./tempIndexTest/typeErrors/" + dir + ".generated.ts"
    );

    try {
      generateTypeScriptFile({
        pathToBnfGrammarFile,
        pathToTypeDictFile,
        pathToOutputFile,
      });
    } catch (e) {
      expect(e).toMatchSnapshot("error for " + dir);
    }

    if (fs.existsSync(pathToOutputFile)) {
      expect(
        getErrorDiagnostics([pathToOutputFile]).length
      ).toBeGreaterThanOrEqual(1);
    }
  });
});
