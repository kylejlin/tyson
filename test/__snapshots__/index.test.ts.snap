// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failure cases: error for missingStartSymbol 1`] = `[Error: The grammar defined in /Users/kyle/tyson/test/fixtures/shouldFail/missingStartSymbol/missingStartSymbol.jison does not have a start symbol declared. Please declare one using \`%start <symbolName>\`.]`;
